File: bogglePy/boggle/partida.py
from boggle.cubilete import Cubilete
import time
import requests

from boggle import cubilete

"""
 * Clase Partida.
 * 
 * Atributos: -maxRondas <-- Almacena el número máximo de rondas que puede tener
 * una partida. -partidasCreadas <-- Almacena un número entero que representa
 * las veces que se han inicializado partidas. -jugadores[] <-- Array de objetos
 * de tipo Jugador. -numRondas <-- Número entero.
 * 
 * Métodos: -iniciar_partida <-- Se encarga de inicializar la partida y gestionar
 * los turnos de los jugadores. -decide_ganador <-- Lee la puntuación de los
 * jugadores de la partida y devuelve el que tenga la puntuación más alta.
 * 
 * @author David Fontalba
 * @version 1.1.
"""

class Partida:
    """
    @param numRondas El número de rondas que se jugará
    @param jugadores Los jugadores que van a participar
    """
    # Atributos de la clase, el número de partidas creadas.
    __MAXRONDAS = 5
    __partidas_creadas = 0

    # Atributos de la partida.
    __jugadores = []
    __num_rondas = -1
    __cubilete = Cubilete()

    def __init__(self, num_rondas, jugadores):
        self.__class__.__partidas_creadas += 1
        self.__num_rondas = min(self.__MAXRONDAS, num_rondas)

        self.__jugadores = jugadores


    def iniciar_partida(self):
        """
        iniciar_partida se encarga del desarrollo de los turnos, y al final llama a {@link decide_ganador}
        """

        # Bucle para las rondas
        for i in range(0, self.__num_rondas):
            print("Prepárate, la ronda " + str(i + 1) + " va a comenzar.")

            # Bucle para los turnos
            for j in self.__jugadores:
                print("Es el turno de " + str(j) + ".")
                # Añadida tirada de dado y muestra del resultado
                self.__cubilete.tirar_dados()
                print(self.__cubilete)
                # Quitamos las repeticiones
                palabras = j.inicio_turno()
                print("\nFin del turno " + str(i + 1) + ".")

                palabras_procesadas = self.comprueba(palabras)

                print("Palabras válidas")
                for palabraProcesada in palabras_procesadas:
                    print(" - " + palabraProcesada)

                j.sumar_puntuacion(self.suma_puntos(palabras_procesadas))

                time.sleep(1)

            print("Fin de la ronda " + str(i + 1) + ".")

        self.decide_ganador()
    def decide_ganador(self):


        ganador = []  # Almacena la posición del jugador que tiene más
        # puntuación
        aux = 0  # Almacena la puntuación máxima

        # Bucle para leer la puntuación de los jugadores
        for (i, j) in enumerate(self.__jugadores):

            # Si el jugador a leer tiene más puntuación
            if j.puntuacion > aux:
                ganador.clear()
                ganador.append(i)
                aux = j.puntuacion

                # Si el jugador a leer tiene la misma puntuación
            elif j.puntuacion == aux:
                ganador.append(i)

        # Si hay más de un ganador
        if (len(ganador) > 1):
            print("Felicidades a los ganadores, ", end="")

            for i in ganador:
                # El último ganador de la lista
                if (i + 1 == len(ganador)):
                    print("y " + self.jugadores[i].nombre + ".")

                    # El resto de ganadores
                else:
                    print(self.jugadores[i].nombre + ", ", end="")
        # Si solo hay un ganador
        else:
            print("Felicidades " + self.jugadores[ganador[0]].nombre + ", ¡Has ganado!")

        print("Recuento de puntuaciones: ")
        for jugador in self.__jugadores:
            print(jugador.nombre + ": " + str(jugador.puntuacion))


    def comprueba(self, a_filtrar):
        """
        comprueba se encarga de comprobar que las palabras sean correctas y filtra
        las incorrectas.
        
        @param aFiltrar La lista de palabras no filtradas
        @return La lista de palabras filtrada
        """
        palabras_filtradas = []

        for palabra_no_filtrada in a_filtrar:

            if (len(palabra_no_filtrada) < 3 or len(palabra_no_filtrada) > 23):
                continue

            palabra_no_filtrada = self.comprobar_existencia_palabra(palabra_no_filtrada)

            if (palabra_no_filtrada == ""):
                continue

            palabra_no_filtrada = self.comprobar_matriz_bien_formada(palabra_no_filtrada)

            if (palabra_no_filtrada == ""):
                continue

            palabras_filtradas.append(palabra_no_filtrada)

        return palabras_filtradas


    def suma_puntos(self, palabras):
        """
        suma_puntos se encarga de sumar los puntos de las palabras
        
        @param palabras La lista de palabras que ya ha sido filtrada
        @return La suma de lo que puntúa cada palabra
        """

        resultado_final = 0

        for palabra in palabras:
            if len(palabra) in [0, 1, 2, 3]:
                pass
            elif len(palabra) == 4:
                resultado_final += 1
            elif len(palabra) == 5:
                resultado_final += 2
            elif len(palabra) == 6:
                resultado_final += 3
            elif len(palabra) == 7:
                resultado_final += 5
            else:
                resultado_final += 11

        return resultado_final
    def comprobar_existencia_palabra(self, palabra_a_filtrar: str):
        """
        comprobar_existencia_palabra comprueba a través de un api la existencia de la palabra
        que llega a través del argumento palabraAFiltrar
        
        @param palabraAFiltrar La palabra a comprobar si existe
        @return Devulve la palabra si existe, en caso contrario, devuelve una cadena vacía
        """

        try:

            url = f"https://dle.rae.es/{palabra_a_filtrar.lower()}"

            conn = requests.get(url)

            exists = f"La palabra <b>{palabra_a_filtrar.lower()}</b> no está en el Diccionario." in conn.text
            if  exists:
                return ""

        except:
            return ""

        return palabra_a_filtrar

    def comprobar_matriz_bien_formada(self, palabra):
        """
        Compruba si la palabra está bien formada respecto a lo que dicen las normas del Boggle,
        es decir que que letra de la palabra tenga que estar continua o adyadcente tanto a la
        siguiente como a la anterior.
        
        @param palabra Palabra para comprobar
        @return Devulve la palabra si está bien formada, en caso contrario, devuelve una {@link String} vacía
        """
        c = self.__cubilete.caras
        palabra = palabra.upper()
        validador = ""

        for x in range(0, 5):
            for y in range(0, 5):
                new_x = x
                new_y = y
                for letra_pos in range(len(palabra)):
                    if c[new_x][new_y] == palabra[letra_pos]:

                        validador += c[new_x][new_y]
                        if validador == palabra:
                            return palabra

                        try:
                            if c[new_x + 1][new_y] == palabra[letra_pos + 1]:
                                new_x = new_x + 1
                                continue

                        except IndexError:
                            pass
                        try:
                            if c[new_x - 1][new_y] == palabra[letra_pos + 1]:
                                new_x = new_x - 1

                        except IndexError:
                            pass
                        try:
                            if c[new_x][new_y + 1] == palabra[letra_pos + 1]:
                                new_y = new_y + 1
                                continue

                        except IndexError:
                            pass
                        try:
                            if c[new_x][new_y - 1] == palabra[letra_pos + 1]:
                                new_y = new_y - 1
                                continue

                        except IndexError:
                            pass
                        try:
                            if c[new_x + 1][new_y + 1] == palabra[letra_pos + 1]:
                                new_y = new_y + 1
                                new_x = new_x + 1
                                continue

                        except IndexError:
                            pass
                        try:
                            if c[new_x + 1][new_y - 1] == palabra[letra_pos + 1]:
                                new_y = new_y - 1
                                new_x = new_x + 1
                                continue

                        except IndexError:
                            pass
                        try:
                            if c[new_x - 1][new_y + 1] == palabra[letra_pos + 1]:
                                new_y = new_y + 1
                                new_x = new_x - 1
                                continue

                        except IndexError:
                            pass
                        try:
                            if c[new_x - 1][new_y - 1] == palabra[letra_pos + 1]:
                                new_y = new_y - 1
                                new_x = new_x - 1
                                continue

                        except IndexError:
                            pass
                    else:
                        validador = ""
                        break

                    letra_pos += 1

        return ""

    # Getters
    def get_partidas_creadas(self):
        """
        @return El número de partidas creadas
        """
        return self.__class__.__partidas_creadas

File: src/boggle/juego/Partida.java
package boggle.juego;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Set;
import javax.swing.JOptionPane;
import boggle.utiles.Teclado;

/**
 * Clase Partida.
 * 
 * Atributos: -maxRondas <-- Almacena el número máximo de rondas que puede tener
 * una partida. -partidasCreadas <-- Almacena un número entero que representa
 * las veces que se han inicializado partidas. -jugadores[] <-- Array de objetos
 * de tipo Jugador. -numRondas <-- Número entero.
 * 
 * Métodos: -iniciarPartida <-- Se encarga de inicializar la partida y gestionar
 * los turnos de los jugadores. -decideGanador <-- Lee la puntuación de los
 * jugadores de la partida y devuelve el que tenga la puntuación más alta.
 * 
 * @author David Fontalba
 * @version 1.1.
 *
 */
public class Partida {
  // Atributos de la clase, el número de partidas creadas.
  private static int MAXRONDAS = 5;
  private static int partidasCreadas = 0;

  // Atributos de la partida.
  public Cubilete cubilete = new Cubilete();
  private Jugador jugadores[];
  private int numRondas;

  /**
   * @param numRondas El número de rondas que se jugará
   * @param jugadores Los jugadores que van a participar
   */
  public Partida(int numRondas, Jugador... jugadores) {
    this.numRondas = Math.min(MAXRONDAS, numRondas);
    this.jugadores = jugadores;
  }

  /**
   * iniciarPartida se encarga del desarrollo de los turnos, y al final llama a {@link decideGanador}
   */
  public void iniciarPartida() {

    // Bucle para las rondas
    for (int i = 0; i < this.numRondas; i++) {
      System.out.println("Prepárate, la ronda " + (i + 1) + " va a comenzar.");

      // Bucle para los turnos
      for (int j = 0; j < this.jugadores.length; j++) {
        System.out.println("Es el turno de " + this.jugadores[j] + ".");
        // Añadida tirada de dado y muestra del resultado
        this.cubilete.tirarDados();
        System.out.println(this.cubilete.toString());
        // Quitamos las repeticiones
        Set<String> palabras = jugadores[j].inicioTurno();
        System.out.println("\nFin del turno " + (j + 1) + ".");

        ArrayList<String> palabrasProcesadas = comprueba(palabras);

        System.out.println("Palabras válidas");
        for (String palabraProcesada : palabrasProcesadas) {
          System.out.println(" - " + palabraProcesada);
        }

        jugadores[j].sumaPuntuacion(sumaPuntos(palabrasProcesadas));

        try {
          Thread.sleep(1000);
        } catch (InterruptedException e) {
        }

      }
      System.out.println("Fin de la ronda " + (i + 1) + ".");

    }

    decideGanador();

  }

  /**
   * decideGanador compara la puntacion de cada jugador y el jugador 
   * que tenga más puntuación gana.
   */
  private void decideGanador() {
    ArrayList<Integer> ganador = new ArrayList<Integer>(); // Almacena la posición del jugador que tiene más
    // puntuación
    int aux = 0; // Almacena la puntuación máxima

    // Bucle para leer la puntuación de los jugadores
    for (int i = 0; i < this.jugadores.length; i++) {

      // Si el jugador a leer tiene más puntuación
      if (this.jugadores[i].getPuntuacion() > aux) {
        ganador.clear();
        ganador.add(i);
        aux = this.jugadores[i].getPuntuacion();

        // Si el jugador a leer tiene la misma puntuación
      } else if (this.jugadores[i].getPuntuacion() == aux) {
        ganador.add(i);
      }

    }

    // Si hay más de un ganador
    if (ganador.size() > 1) {
      System.out.print("Felicidades a los ganadores, ");

      for (int i = 0; i < ganador.size(); i++) {
        // El último ganador de la lista
        if (i + 1 == ganador.size()) {
          System.out.println("y " + this.jugadores[ganador.get(i)].getNombre() + ".");

          // El resto de ganadores
        } else {
          System.out.print(this.jugadores[ganador.get(i)].getNombre() + ", ");
        }

      }
      // Si solo hay un ganador
    } else {
      System.out.println("Felicidades " + this.jugadores[ganador.get(0)].getNombre() + ", ¡Has ganado!");

    }

    System.out.println("Recuento de puntuaciones: ");
    for (Jugador jugador : this.jugadores) {
      System.out.println(jugador.getNombre() + ": " + jugador.getPuntuacion());
    }
  }

  /**
   * comprueba se encarga de comprobar que las palabras sean correctas y filtra
   * las incorrectas.
   * 
   * @param aFiltrar La lista de palabras no filtradas
   * @return La lista de palabras filtrada
   */
  private ArrayList<String> comprueba(Set<String> aFiltrar) {

    ArrayList<String> palabrasFiltradas = new ArrayList<>();

    for (String palabraNoFiltrada : aFiltrar) {

      if (palabraNoFiltrada.length() < 3 && palabraNoFiltrada.length() > 23) {
        continue;
      }

      palabraNoFiltrada = comprobarExistenciaPalabra(palabraNoFiltrada);

      if (palabraNoFiltrada.isEmpty()) {
        continue;
      }
      palabraNoFiltrada = comprobarMatrizBienFormada(palabraNoFiltrada);

      if (palabraNoFiltrada.isEmpty()) {
        continue;
      }

      palabrasFiltradas.add(palabraNoFiltrada);

    }

    return palabrasFiltradas;
  }

  /**
   * 
   * sumaPuntos se encarga de sumar los puntos de las palabras
   * 
   * @param palabras La lista de palabras que ya ha sido filtrada
   * @return La suma de lo que puntúa cada palabra
   */
  private int sumaPuntos(ArrayList<String> palabras) {

    int resultadoFinal = 0;

    for (String palabra : palabras) {
      switch (palabra.length()) {
        case 0:
        case 1:
        case 2:
        case 3:
          break;
        case 4:
          resultadoFinal += 1;
          break;
        case 5:
          resultadoFinal += 2;
          break;
        case 6:
          resultadoFinal += 3;
          break;
        case 7:
          resultadoFinal += 5;
          break;
        default:
          resultadoFinal += 11;
          break;
      }
    }

    return resultadoFinal;
  }

  /**
   * 
   * comprobarExistenciaPalabra comprueba a través de un api la existencia de la palabra
   * que llega a través del argumento palabraAFiltrar
   * 
   * @param palabraAFiltrar La palabra a comprobar si existe
   * @return Devulve la palabra si existe, en caso contrario, devuelve una {@link String} vacía
   */
  private String comprobarExistenciaPalabra(String palabraAFiltrar) {

    try {

      URL url = new URL(
          String.format("https://od-api.oxforddictionaries.com:443/api/v2/entries/es/%s?lexicalCategory=noun,verb",
              palabraAFiltrar.toLowerCase()));
      HttpURLConnection conn = (HttpURLConnection) url.openConnection();
      conn.setRequestMethod("GET");
      conn.setRequestProperty("app_id", "46a14e95");
      conn.setRequestProperty("app_key", "384d633d7a131cd0ab86fb6322665e15");
      conn.connect();

      if (conn.getResponseCode() != 200) {
        return "";
      }

    } catch (MalformedURLException e) {
      return "";
    } catch (IOException e) {
      return "";
    }

    return palabraAFiltrar;

  }

  /**
   * 
   * Compruba si la palabra está bien formada respecto a lo que dicen las normas del Boggle,
   * es decir que que letra de la palabra tenga que estar continua o adyadcente tanto a la
   * siguiente como a la anterior.
   * 
   * @param palabra Palabra para comprobar
   * @return Devulve la palabra si está bien formada, en caso contrario, devuelve una {@link String} vacía
   */
  private String comprobarMatrizBienFormada(String palabra) {
    char[][] c = cubilete.caras;
    palabra = palabra.toUpperCase();
    String validador = "";

    for (int x = 0; x < 5; x++) {
      for (int y = 0; y < 5; y++) {
        int newX = x;
        int newY = y;
        for (int letraPos = 0; letraPos < palabra.length(); letraPos++) {
          if (c[newX][newY] == (palabra.charAt(letraPos))) {

            validador += c[newX][newY];
            if (validador.equals(palabra)) {
              return palabra;
            }
            try {
              if (c[newX + 1][newY] == (palabra.charAt(letraPos + 1))) {
                newX = newX + 1;
                continue;
              }
            } catch (IndexOutOfBoundsException e) {
            }
            try {
              if (c[newX - 1][newY] == (palabra.charAt(letraPos + 1))) {
                newX = newX - 1;
              }
            } catch (IndexOutOfBoundsException e) {
            }
            try {
              if (c[newX][newY + 1] == (palabra.charAt(letraPos + 1))) {
                newY = newY + 1;
                continue;
              }
            } catch (IndexOutOfBoundsException e) {
            }
            try {
              if (c[newX][newY - 1] == (palabra.charAt(letraPos + 1))) {
                newY = newY - 1;
                continue;
              }
            } catch (IndexOutOfBoundsException e) {
            }
            try {
              if (c[newX + 1][newY + 1] == (palabra.charAt(letraPos + 1))) {
                newY = newY + 1;
                newX = newX + 1;
                continue;
              }
            } catch (IndexOutOfBoundsException e) {
            }
            try {
              if (c[newX + 1][newY - 1] == (palabra.charAt(letraPos + 1))) {
                newY = newY - 1;
                newX = newX + 1;
                continue;
              }
            } catch (IndexOutOfBoundsException e) {
            }
            try {
              if (c[newX - 1][newY + 1] == (palabra.charAt(letraPos + 1))) {
                newY = newY + 1;
                newX = newX - 1;
                continue;
              }
            } catch (IndexOutOfBoundsException e) {
            }
            try {
              if (c[newX - 1][newY - 1] == (palabra.charAt(letraPos + 1))) {
                newY = newY - 1;
                newX = newX - 1;
                continue;
              }
            } catch (IndexOutOfBoundsException e) {
            }
          } else {
            validador = "";
            break;
          }
          letraPos++;

        }

      }
    }

    return "";

  }

  /**
   * @return El número de partidas creadas
   */
  public static int getPartidasCreadas() {
    return partidasCreadas;
  }
}